pty_windows.go

```diff
@@ -0,0 +1,160 @@
+package pty
+
+import (
+	"os"
+	"syscall"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+)
+
+var (
+	kernel32DLL *syscall.DLL
+
+	createPseudoConsole        *syscall.Proc
+	resizePseudoConsole        *syscall.Proc
+	closePseudoConsole         *syscall.Proc
+	getConsoleScreenBufferInfo *syscall.Proc
+
+	initializeProcThreadAttributeList *syscall.Proc
+	updateProcThreadAttribute         *syscall.Proc
+	deleteProcThreadAttributeList     *syscall.Proc
+)
+
+func init() {
+	var err error
+	kernel32DLL, err = syscall.LoadDLL("kernel32.dll")
+	if err != nil {
+		panic(err)
+	}
+
+	createPseudoConsole = kernel32DLL.MustFindProc("CreatePseudoConsole")
+	resizePseudoConsole = kernel32DLL.MustFindProc("ResizePseudoConsole")
+	closePseudoConsole = kernel32DLL.MustFindProc("ClosePseudoConsole")
+	getConsoleScreenBufferInfo = kernel32DLL.MustFindProc("GetConsoleScreenBufferInfo")
+
+	initializeProcThreadAttributeList = kernel32DLL.MustFindProc("InitializeProcThreadAttributeList")
+	updateProcThreadAttribute = kernel32DLL.MustFindProc("UpdateProcThreadAttribute")
+	deleteProcThreadAttributeList = kernel32DLL.MustFindProc("DeleteProcThreadAttributeList")
+}
+
+func open() (_ Pty, _ Tty, err error) {
+	pr, consoleW, err := os.Pipe()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	consoleR, pw, err := os.Pipe()
+	if err != nil {
+		_ = consoleW.Close()
+		_ = pr.Close()
+		return nil, nil, err
+	}
+
+	defer func() {
+		if err != nil {
+			_ = consoleW.Close()
+			_ = pr.Close()
+
+			_ = pw.Close()
+			_ = consoleR.Close()
+		}
+	}()
+
+	var (
+		consoleHandle syscall.Handle
+		defaultSize   = &windows.Coord{X: 80, Y: 30}
+	)
+	// https://docs.microsoft.com/en-us/windows/console/createpseudoconsole
+	r1, _, err := createPseudoConsole.Call(
+		uintptr(unsafe.Pointer(defaultSize)),    // size: default 80x30 window
+		consoleR.Fd(),                           // console input
+		consoleW.Fd(),                           // console output
+		0,                                       // console flags, currently only PSEUDOCONSOLE_INHERIT_CURSOR supported
+		uintptr(unsafe.Pointer(&consoleHandle)), // console handler value return
+	)
+	if r1 != 0 {
+		// S_OK: 0
+		return nil, nil, os.NewSyscallError("CreatePseudoConsole", err)
+	}
+
+	return &WindowsPty{
+			handle:    uintptr(consoleHandle),
+			r:         pr,
+			w:         pw,
+			_consoleR: consoleR,
+			_consoleW: consoleW,
+		}, &WindowsTty{
+			handle: uintptr(consoleHandle),
+			r:      consoleR,
+			w:      consoleW,
+		}, nil
+}
+
+var _ Pty = (*WindowsPty)(nil)
+
+type WindowsPty struct {
+	handle uintptr
+	r, w   *os.File
+
+	_consoleR, _consoleW *os.File
```

Comment:
nit: please avoid _ prefix. those are already private.
```suggestion
	consoleR, consoleW *os.File
```

Reply:
I see it as a special way to catch developer's attention, what do you think?

Reply:
It is similar to ptmx/pts in unix where the slave is used by the child process but managed by us.